Prefer python meta function over c++ meta function

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87662
* #87563
* #87569
* __->__ #87426

This is a policy update for meta registration. **We now prefer python meta implementation over C++ meta function.**  This is a flip of the previous policy, where we prefer C++ meta function over python meta function if they both exist. 

Here's the meta registration process:
1. register_meta and register_decomposition will place the python meta/decomp functions into the `global_decomp_table`.  However, they will NOT register them into dispatcher. 
2. After global_decomp_table is populated, we will compile an `active_meta_table`. For a given op, we pick the most specific decomp function from `global_decomp_table` in the preference order of Meta > PostAutograd > PreAutograd. 
3. We will unconditionally register all of them into python dispatcher. And register them into C++ dispatcher, unless it one of the following 3 cases
- 1. the op is a CompositeImplicitAutograd, and should rely on decomposed op's meta
- 2. the op is a view op, as the MetaTensor doesn't support aliased storage
- 3. the op is in the blocklist (due to UT failures, and we will burn down this list op by op) 

Over the long run, we wish to implement all meta functions in python. With this PR, 321 op_overloads will have cpp meta overridden by python meta. There are still 400 op_overloads is using cpp meta. The exact list can be found here https://gist.github.com/SherlockNoMad/d20bb736178df8eebd3b054c8bb7cdc5



cc @ngimel @jansel @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @lezcano @fdrocha

SherlockNoMad committed:
Unconditionally override c++ meta function

Remove disable_meta flag

Remove register_dispatcher flag

[ghstack-poisoned]

eellison commented on torch/_meta_registrations.py:
```diff
@@ -1636,3 +1638,64 @@ def upsample_nearest2d_vec(input, output_size, scale_factors):
+            "aten::mul.Tensor",  # ValueError: Receive two Number inputs to an elementwise binary operation! test/inductor/test_torchinductor.py -k test_both_scalars
+            "aten::div.Tensor",  # ValueError: Receive two Number inputs to an elementwise binary operation! test/test_fake_tensor.py -k test_scalar_inputs
+            "aten::div.Tensor_mode",  # ValueError: Receive two Number inputs to an elementwise binary operation! test/inductor/test_torchinductor.py -k test_div8_cpu
+            "aten::diag_embed",  # RuntimeError: Stride mismatch! Strides are (180, 30, 1, 6) and (180, 30, 5, 1) (mismatched at 2)! test/test_ops.py -k test_fake_autocast_diag_embed_cuda_float32
```
it seems like we're just moving the public api into a private list. Is that useful ?

SherlockNoMad replied:
This is a list I will work to burn down. 
Most of them can be resolved by fixing the meta implementation. 

For the ones that python meta doesn't work, I plan to add it to the block list in functions_with_cpp_meta_impl_that_support_symint. 

So eventually, it should just look up from functions_with_cpp_meta_impl_that_support_symint.

eellison reviewed (commented):
Couple comments/questions

SherlockNoMad commented on torch/_decomp/decompositions.py:
```diff
@@ -1929,7 +1929,7 @@ def is_same_size(a: Tensor, b: Tensor) -> bool:
-@register_decomposition([aten._reshape_alias, aten._unsafe_view], disable_meta=True)
+@register_decomposition([aten._reshape_alias, aten._unsafe_view])
 def _reshape_alias(x, shape, *args):
     return aten.view(x, shape)
```
_reshape_alias is not registered to C++ dispatcher, as it's a view op. See the registration rule in the description.

But yes, we should still fix the stride. I am wondering if we can track this by asserting the stride correctness in test_meta.py.

Chillee replied:
I don't see why that's incorrect? From my understanding, `_reshape_alias` is just `view` but with the strides precomputed.

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorShape.cpp#L1419

<img width="528" alt="image" src="https://user-images.githubusercontent.com/6355099/197085092-11cee693-0729-425b-9d65-44f061af84d0.png">

eellison replied:
https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/native/TensorShape.cpp.html#_ZN2at6native14_reshape_aliasERKNS_6TensorEN3c108ArrayRefIlEES6_


```
>>> x = torch.rand([8, 4])
>>> y = torch.ops.aten._reshape_alias(x, [8, 4], [1, 4])
>>> y.stride()
(1, 4)
>>> x.view([8, 4]).stride()
(4, 1)
>>> 
```

If the strides are always pre-computed to be consistent we should assert that.

SherlockNoMad commented on torch/_meta_registrations.py:
```diff
@@ -1636,3 +1638,64 @@ def upsample_nearest2d_vec(input, output_size, scale_factors):
+    for op_overload, fn in activate_meta_table.items():
+        assert isinstance(op_overload, OpOverload)
+
+        op_overload.py_impl(torch._C.DispatchKey.Meta)(fn)
```
@ezyang, 
I am wondering if we should make python dispatcher following the registration logic as c++ dispatcher. 

If yes, we will have the consistent behavior between python dispatcher and c++ dispatcher on Meta key.

SherlockNoMad replied:
OK, will do in next PR.

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -147,13 +147,15 @@ OperatorEntry::AnnotatedKernelContainerIterator OperatorEntry::registerKernel(
-               "  previous kernel: ", (cpp_signature_.has_value() ? cpp_signature_->debug : (sym_cpp_signature_.has_value() ? sym_cpp_signature_->debug : "no debug info")), "\n",
-               "       new kernel: ", debug
-    );
+    if (dispatch_key != DispatchKey::Meta) {
```
Good to have a comment why Meta is suppressed specifically

SherlockNoMad replied:
done.

SherlockNoMad committed:
Update on "Prefer python meta function over c++ meta function"


This is a policy update for meta registration. **We now prefer python meta implementation over C++ meta function.**  This is a flip of the previous policy, where we prefer C++ meta function over python meta function if they both exist. 

Here's the meta registration process:
1. register_meta and register_decomposition will place the python meta/decomp functions into the `global_decomp_table`.  However, they will NOT register them into dispatcher. 
2. After global_decomp_table is populated, we will compile an `active_meta_table`. For a given op, we pick the most specific decomp function from `global_decomp_table` in the preference order of Meta > PostAutograd > PreAutograd. 
3. We will unconditionally register all of them into python dispatcher. And register them into C++ dispatcher, unless it one of the following 3 cases
- 1. the op is a CompositeImplicitAutograd, and should rely on decomposed op's meta
- 2. the op is a view op, as the MetaTensor doesn't support aliased storage
- 3. the op is in the blocklist (due to UT failures, and we will burn down this list op by op) 

Over the long run, we wish to implement all meta functions in python. With this PR, 321 op_overloads will have cpp meta overridden by python meta. There are still 400 op_overloads is using cpp meta. The exact list can be found here https://gist.github.com/SherlockNoMad/d20bb736178df8eebd3b054c8bb7cdc5



cc ngimel jansel lezcano fdrocha mlazos soumith

[ghstack-poisoned]

SherlockNoMad commented on test/test_dispatch.py:
```diff
@@ -765,6 +765,7 @@ def test_overwrite_math(self):
     # no longer exists
     def test_find_dangling_impls(self):
         dangling_impls = C._dispatch_find_dangling_impls()
+        print(dangling_impls)
```
removed.

ezyang commented on test/test_functionalization.py:
```diff
@@ -1033,8 +1033,8 @@ def forward(self, a_1):
-    fill = torch.ops.aten.fill_.Scalar(view, 1)
-    view_1 = torch.ops.aten.view.default(view, [5, 5]);  view = None
+    fill = torch.ops.aten.fill.Scalar(view, 1);  view = None
+    view_1 = torch.ops.aten.view.default(fill, [5, 5]);  fill = None
```
This is a regression , cc @bdhirsh

bdhirsh replied:
probably a bug in one of our (related) python decomps?

SherlockNoMad replied:
fixed. 
It's a prim's implementation bug.

SherlockNoMad replied:
Fixed here https://github.com/pytorch/pytorch/pull/87493

ezyang commented on test/test_ops.py:
```diff
@@ -1909,6 +1908,8 @@ def test_refs_are_in_decomp_table(self, op):
     skip("linalg.pinv", "hermitian"),
     skip("linalg.pinv", "singular"),
     skip('pinverse'),
+    skip('cholesky_inverse'),  # test/test_ops.py -k test_fake_crossref_backward_amp_cholesky_inverse_cuda_float32
```
this failing now is weird

SherlockNoMad replied:
Fixed. It was flaky, and not failing now...

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -780,7 +780,6 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         if (
             func in decomposition_table
             and torch_decomp_decompositions(func)
-            and func not in _disabled_meta_decomps
```
@eellison check this

eellison replied:
@SherlockNoMad, can you find to expose the set of ops you've disabled for meta decomps to not be called ? (previously `_disabled_meta_decomps`) ?

SherlockNoMad replied:
Here's the exhausted list. Essentially, all the decomps previous marked with "disable_meta=True" will be in this list. 

```
aten::split_with_sizes
aten::split.Tensor
aten::native_group_norm
aten::detach
aten::lift
aten::lift.out
aten::lift_fresh
aten::_adaptive_avg_pool2d
aten::_adaptive_avg_pool2d.out
aten::_reshape_alias
aten::_unsafe_view
aten::_unsafe_view.out
aten::dot
aten::dot.out
aten::expand
aten::permute
aten::index_select
aten::index_select.out
aten::index_select.dimname
aten::index_select.dimname_out
aten::squeeze
aten::squeeze.dim
aten::squeeze.dimname
aten::diagonal
aten::diagonal.Dimname
aten::t
aten::transpose.int
aten::transpose.Dimname
aten::unsqueeze
aten::view
aten::view.dtype
```

eellison replied:
Yea, we can we avoid calling those decopmositions here in fake tensor, as we were before ?

ezyang commented on torch/_meta_registrations.py:
```diff
@@ -1636,3 +1629,60 @@ def upsample_nearest2d_vec(input, output_size, scale_factors):
+    activate_meta_table = {}
+
+    for type in ["meta", "post_autograd", "pre_autograd"]:
+        registry = global_decomposition_table[type]
```
why not just iterate over the keys of the table

ezyang replied:
oh nvm

ezyang commented on torch/_meta_registrations.py:
```diff
@@ -1636,3 +1629,60 @@ def upsample_nearest2d_vec(input, output_size, scale_factors):
+
+    activate_meta_table = {}
+
+    for type in ["meta", "post_autograd", "pre_autograd"]:
```
comment something about precedence order

SherlockNoMad replied:
done.

ezyang commented on torch/_meta_registrations.py:
```diff
@@ -1636,3 +1629,60 @@ def upsample_nearest2d_vec(input, output_size, scale_factors):
+
+        op_overload.py_impl(torch._C.DispatchKey.Meta)(fn)
+
+        dispatch_key_registration = torch._C._dispatch_dump(op_overload.name())
```
don't do it this way; use one of the more specific dispatch query functions

SherlockNoMad replied:
Shall I also check for CompositeImplicitAutogradNestedTensor? 

aten.unbind has this Dispatchkey.

SherlockNoMad replied:
I suppose it's not needed? 

For non-nested tensor, aten.unbind should still depends on python's meta.

SherlockNoMad replied:
uhhh... library.py is checking CompositeImplicitAutogradNestedTensor


I am fix this in the next PR.

ezyang replied:
for the CompositeImplicitAutograd test you just check if the (non-computed) key is registered

ezyang reviewed (approved):
some small nits, looks good, much simpler

SherlockNoMad committed:
Update on "Prefer python meta function over c++ meta function"


This is a policy update for meta registration. **We now prefer python meta implementation over C++ meta function.**  This is a flip of the previous policy, where we prefer C++ meta function over python meta function if they both exist. 

Here's the meta registration process:
1. register_meta and register_decomposition will place the python meta/decomp functions into the `global_decomp_table`.  However, they will NOT register them into dispatcher. 
2. After global_decomp_table is populated, we will compile an `active_meta_table`. For a given op, we pick the most specific decomp function from `global_decomp_table` in the preference order of Meta > PostAutograd > PreAutograd. 
3. We will unconditionally register all of them into python dispatcher. And register them into C++ dispatcher, unless it one of the following 3 cases
- 1. the op is a CompositeImplicitAutograd, and should rely on decomposed op's meta
- 2. the op is a view op, as the MetaTensor doesn't support aliased storage
- 3. the op is in the blocklist (due to UT failures, and we will burn down this list op by op) 

Over the long run, we wish to implement all meta functions in python. With this PR, 321 op_overloads will have cpp meta overridden by python meta. There are still 400 op_overloads is using cpp meta. The exact list can be found here https://gist.github.com/SherlockNoMad/d20bb736178df8eebd3b054c8bb7cdc5



cc ngimel jansel lezcano fdrocha mlazos soumith

[ghstack-poisoned]

eellison commented on test/test_ops.py:
```diff
@@ -1909,6 +1908,7 @@ def test_refs_are_in_decomp_table(self, op):
     skip("linalg.pinv", "hermitian"),
     skip("linalg.pinv", "singular"),
     skip('pinverse'),
+    skip('cholesky_inverse'),  # test/test_ops.py -k test_fake_crossref_backward_amp_cholesky_inverse_cuda_float32
```
Would prefer if we didn't regress the tests to land this - could you look into this ?

SherlockNoMad replied:
Fixed. It was flaky, and not failing now...

SherlockNoMad committed:
Update on "Prefer python meta function over c++ meta function"


This is a policy update for meta registration. **We now prefer python meta implementation over C++ meta function.**  This is a flip of the previous policy, where we prefer C++ meta function over python meta function if they both exist. 

Here's the meta registration process:
1. register_meta and register_decomposition will place the python meta/decomp functions into the `global_decomp_table`.  However, they will NOT register them into dispatcher. 
2. After global_decomp_table is populated, we will compile an `active_meta_table`. For a given op, we pick the most specific decomp function from `global_decomp_table` in the preference order of Meta > PostAutograd > PreAutograd. 
3. We will unconditionally register all of them into python dispatcher. And register them into C++ dispatcher, unless it one of the following 3 cases
- 1. the op is a CompositeImplicitAutograd, and should rely on decomposed op's meta
- 2. the op is a view op, as the MetaTensor doesn't support aliased storage
- 3. the op is in the blocklist (due to UT failures, and we will burn down this list op by op) 

Over the long run, we wish to implement all meta functions in python. With this PR, 321 op_overloads will have cpp meta overridden by python meta. There are still 400 op_overloads is using cpp meta. The exact list can be found here https://gist.github.com/SherlockNoMad/d20bb736178df8eebd3b054c8bb7cdc5



cc ngimel jansel lezcano fdrocha mlazos soumith

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Prefer python meta function over c++ meta function"


This is a policy update for meta registration. **We now prefer python meta implementation over C++ meta function.**  This is a flip of the previous policy, where we prefer C++ meta function over python meta function if they both exist. 

Here's the meta registration process:
1. register_meta and register_decomposition will place the python meta/decomp functions into the `global_decomp_table`.  However, they will NOT register them into dispatcher. 
2. After global_decomp_table is populated, we will compile an `active_meta_table`. For a given op, we pick the most specific decomp function from `global_decomp_table` in the preference order of Meta > PostAutograd > PreAutograd. 
3. We will unconditionally register all of them into python dispatcher. And register them into C++ dispatcher, unless it one of the following 3 cases
- 1. the op is a CompositeImplicitAutograd, and should rely on decomposed op's meta
- 2. the op is a view op, as the MetaTensor doesn't support aliased storage
- 3. the op is in the blocklist (due to UT failures, and we will burn down this list op by op) 

Over the long run, we wish to implement all meta functions in python. With this PR, 321 op_overloads will have cpp meta overridden by python meta. There are still 400 op_overloads is using cpp meta. The exact list can be found here https://gist.github.com/SherlockNoMad/d20bb736178df8eebd3b054c8bb7cdc5



cc ngimel jansel lezcano fdrocha mlazos soumith voznesenskym yanboliang

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Prefer python meta function over c++ meta function"


This is a policy update for meta registration. **We now prefer python meta implementation over C++ meta function.**  This is a flip of the previous policy, where we prefer C++ meta function over python meta function if they both exist. 

Here's the meta registration process:
1. register_meta and register_decomposition will place the python meta/decomp functions into the `global_decomp_table`.  However, they will NOT register them into dispatcher. 
2. After global_decomp_table is populated, we will compile an `active_meta_table`. For a given op, we pick the most specific decomp function from `global_decomp_table` in the preference order of Meta > PostAutograd > PreAutograd. 
3. We will unconditionally register all of them into python dispatcher. And register them into C++ dispatcher, unless it one of the following 3 cases
- 1. the op is a CompositeImplicitAutograd, and should rely on decomposed op's meta
- 2. the op is a view op, as the MetaTensor doesn't support aliased storage
- 3. the op is in the blocklist (due to UT failures, and we will burn down this list op by op) 

Over the long run, we wish to implement all meta functions in python. With this PR, 321 op_overloads will have cpp meta overridden by python meta. There are still 400 op_overloads is using cpp meta. The exact list can be found here https://gist.github.com/SherlockNoMad/d20bb736178df8eebd3b054c8bb7cdc5



cc ngimel jansel lezcano fdrocha mlazos soumith voznesenskym yanboliang

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Prefer python meta function over c++ meta function"


This is a policy update for meta registration. **We now prefer python meta implementation over C++ meta function.**  This is a flip of the previous policy, where we prefer C++ meta function over python meta function if they both exist. 

Here's the meta registration process:
1. register_meta and register_decomposition will place the python meta/decomp functions into the `global_decomp_table`.  However, they will NOT register them into dispatcher. 
2. After global_decomp_table is populated, we will compile an `active_meta_table`. For a given op, we pick the most specific decomp function from `global_decomp_table` in the preference order of Meta > PostAutograd > PreAutograd. 
3. We will unconditionally register all of them into python dispatcher. And register them into C++ dispatcher, unless it one of the following 3 cases
- 1. the op is a CompositeImplicitAutograd, and should rely on decomposed op's meta
- 2. the op is a view op, as the MetaTensor doesn't support aliased storage
- 3. the op is in the blocklist (due to UT failures, and we will burn down this list op by op) 

Over the long run, we wish to implement all meta functions in python. With this PR, 321 op_overloads will have cpp meta overridden by python meta. There are still 400 op_overloads is using cpp meta. The exact list can be found here https://gist.github.com/SherlockNoMad/d20bb736178df8eebd3b054c8bb7cdc5



cc ngimel jansel lezcano fdrocha mlazos soumith voznesenskym yanboliang

[ghstack-poisoned]

SherlockNoMad commented:
@pytorchbot merge -g

SherlockNoMad committed:
Update on "Prefer python meta function over c++ meta function"


This is a policy update for meta registration. **We now prefer python meta implementation over C++ meta function.**  This is a flip of the previous policy, where we prefer C++ meta function over python meta function if they both exist. 

Here's the meta registration process:
1. register_meta and register_decomposition will place the python meta/decomp functions into the `global_decomp_table`.  However, they will NOT register them into dispatcher. 
2. After global_decomp_table is populated, we will compile an `active_meta_table`. For a given op, we pick the most specific decomp function from `global_decomp_table` in the preference order of Meta > PostAutograd > PreAutograd. 
3. We will unconditionally register all of them into python dispatcher. And register them into C++ dispatcher, unless it one of the following 3 cases
- 1. the op is a CompositeImplicitAutograd, and should rely on decomposed op's meta
- 2. the op is a view op, as the MetaTensor doesn't support aliased storage
- 3. the op is in the blocklist (due to UT failures, and we will burn down this list op by op) 

Over the long run, we wish to implement all meta functions in python. With this PR, 321 op_overloads will have cpp meta overridden by python meta. There are still 400 op_overloads is using cpp meta. The exact list can be found here https://gist.github.com/SherlockNoMad/d20bb736178df8eebd3b054c8bb7cdc5



cc ngimel jansel lezcano fdrocha mlazos soumith voznesenskym yanboliang

[ghstack-poisoned]

SherlockNoMad commented:
@pytorchbot merge -g
